### DISCUSSION TITLE
[Tracking]: Reduce duplications in `nargo_cli` and get wasm-friendly parts

### DISCUSSION THREAD

**Discussion by dmvict** [noir-lang/noir#3704] [Ideas]
### Problem

I made some work with repo

The issue consists of two parts. 

===

First part is duplications in `nargo_cli` `cmd`'s  modules. I noticed that most of the `cmd`'s module  `run` methods have code:
```rs
    let toml_path = get_package_manifest(&config.program_dir)?;
    let default_selection =
        if args.workspace { PackageSelection::All } else { PackageSelection::DefaultOrAll };
    let selection = args.package.map_or(default_selection, PackageSelection::Selected);
    let workspace = resolve_workspace_from_toml(
        &toml_path,
        selection,
        Some(NOIR_ARTIFACT_VERSION_STRING.to_string()),
    )?;  
```
It seems like it can be moved to some init routine. 

===

The second part is related to `wasm` crate. There are some duplication between `wasm` and `nargo_cli` compile module. It is necessary because wasm cannot work with filesystem. See the strategies in `Happy Case` section. 

### Happy Case

Fist part:
- move duplicated code into some init routine 
- alternatively redesign `run` method and its call

Second part:
- generalize compiling algorithms in wasm and nargo_cli, move the code in reusable library. I mean such parts of the code as https://github.com/noir-lang/noir/blob/master/compiler/wasm/src/compile.rs#L167 and https://github.com/noir-lang/noir/blob/master/tooling/nargo_cli/src/cli/compile_cmd.rs#L239 
- update code in related crates
- implement simple wrapper structure for wasm compiler. It should improve work on TS side
- improve work with errors. I see that module implements custom error type for typescript. I think that it is better to use native errors. I can be wrong about it.

Feel free to comment the considerations!

### Would you like to submit a PR for this Issue?

No

**Comment by Savio-Sou** [comment]
cc @kevaundray
